### ISSUE TITLE
NullPointerException in FieldItemSummary 

### ISSUE THREAD

**Issue by romani** [spotbugs/spotbugs#2759]
detected during version bump spotbugs-maven-plugin from 4.7.3.5 to 4.8.2.0
https://github.com/checkstyle/checkstyle/pull/14117

https://app.circleci.com/pipelines/github/checkstyle/checkstyle/22221/workflows/e397e60b-04ca-4618-b94e-5ca0be2a20d8/jobs/446175

OS:
```
 Kernel Version: 5.15.0-1039-aws
 Operating System: Ubuntu 20.04.6 LTS
 OSType: linux
 Architecture: x86_64
``` 

Special docker [container](https://github.com/checkstyle/contribution/blob/master/devops/docker/no-exception-build-image/jdk-11-groovy-git-mvn/Dockerfile):
```
Starting container checkstyle/jdk-11-groovy-git-mvn:11.0.20.1__2.4.21__2.42.0__3.9.5
this means
jdk11.0.20.1
mvn 3.9.5
```

Execution log:
```
.....
[[1;34mINFO[m] [1m--- [0;32mspotbugs:4.8.2.0:spotbugs[m [1m(spotbugs)[m @ [36mcheckstyle[0;1m ---[m
[[1;34mINFO[m] Fork Value is true
     [java] The following errors occurred during analysis:
     [java]   Exception analyzing com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser$ClassBlockContext using detector edu.umd.cs.findbugs.detect.FieldItemSummary
     [java]     java.lang.NullPointerException
     [java]       At edu.umd.cs.findbugs.visitclass.DismantleBytecode.visit(DismantleBytecode.java:719)
     [java]       At edu.umd.cs.findbugs.detect.FieldItemSummary.visit(FieldItemSummary.java:131)
     [java]       At edu.umd.cs.findbugs.visitclass.BetterVisitor.visitCode(BetterVisitor.java:218)
     [java]       At edu.umd.cs.findbugs.visitclass.PreorderVisitor.visitCode(PreorderVisitor.java:250)
     [java]       At edu.umd.cs.findbugs.bcel.OpcodeStackDetector.visitCode(OpcodeStackDetector.java:65)
     [java]       At org.apache.bcel.classfile.Code.accept(Code.java:121)
     [java]       At edu.umd.cs.findbugs.visitclass.PreorderVisitor.doVisitMethod(PreorderVisitor.java:322)
     [java]       At edu.umd.cs.findbugs.visitclass.PreorderVisitor.visitJavaClass(PreorderVisitor.java:404)
     [java]       At org.apache.bcel.classfile.JavaClass.accept(JavaClass.java:244)
     [java]       At edu.umd.cs.findbugs.BytecodeScanningDetector.visitClassContext(BytecodeScanningDetector.java:38)
     [java]       At edu.umd.cs.findbugs.DetectorToDetector2Adapter.visitClass(DetectorToDetector2Adapter.java:76)
     [java]       At edu.umd.cs.findbugs.FindBugs2.lambda$analyzeApplication$1(FindBugs2.java:1108)
     [java]       At java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
     [java]       At edu.umd.cs.findbugs.CurrentThreadExecutorService.execute(CurrentThreadExecutorService.java:86)
     [java]       At java.base/java.util.concurrent.AbstractExecutorService.invokeAll(AbstractExecutorService.java:242)
     [java]       At edu.umd.cs.findbugs.FindBugs2.analyzeApplication(FindBugs2.java:1118)
     [java]       At edu.umd.cs.findbugs.FindBugs2.execute(FindBugs2.java:309)
     [java]       At edu.umd.cs.findbugs.FindBugs.runMain(FindBugs.java:395)
     [java]       At edu.umd.cs.findbugs.FindBugs2.main(FindBugs2.java:1231)
```

**Note:**
on versions below it works wihtou exception (my local):
```
$ mvn --version
Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537)
Maven home: /opt/maven/apache-maven-3.8.4
Java version: 11.0.14.1, vendor: Amazon.com Inc., runtime: /opt/jvm/amazon-corretto-11.0.14.10.1-linux-x64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.4.0-210-generic", arch: "amd64", family: "unix"
```

**Comment by romani** [comment]
the same NPE even on old container that we used for several years "checkstyle/jdk-11-groovy-git-mvn:11.0.13__3.0.9__2.25.1__3.6.3"
so regression is FieldItemSummary

Command that we use in CI for our repo:
`mvn -e --no-transfer-progress clean test-compile pmd:check spotbugs:check`

**Comment by gtoison** [comment]
As far as I can see the code (`edu.umd.cs.findbugs.visitclass.DismantleBytecode` or `edu.umd.cs.findbugs.detect.FieldItemSummary`) hasn't changed much recently so this is very strange indeed.
Is there a way to reproduce the problem without docker (simply with maven?), that would facilitate the troubleshooting

**Comment by gtoison** [comment]
Would it be possible to share a zip with the compiled .class files of the project?

**Comment by romani** [comment]
Here are all binaries https://app.circleci.com/pipelines/github/checkstyle/checkstyle/22599/workflows/002e52f5-5903-49da-be98-1100895997dc/jobs/462597/artifacts, zip is a bit more complicated to make, let's try this way first.

Binaries are generated by pull request https://github.com/checkstyle/checkstyle/pull/14205

In case such PR be closed, we can always reopen it to rerun CI.

**Comment by gtoison** [comment]
Thank you for the samples! I tried to reproduce the problem but strangely got an other error which ultimately boils down to [BCEL-370](https://issues.apache.org/jira/browse/BCEL-370) so I submitted an fix attempt here: https://github.com/apache/commons-bcel/pull/254

Unless I'm mistaken your build runs SpotBugs _after_ Jacoco has modified the class files. The updated bytecode contains instructions that are valid but not normally emitted by compilers. This happens on JDK 11 because the `Constant_Dynamic` instruction was added there, so Jacoco only uses it on >= 11.

So until the issue is fixed in BCEL (and maybe SpotBugs) I think that you should try a find a way to run SpotBugs before Jacoco

**Comment by gtoison** [comment]
So this should be fixed in the upcoming released now the problem in BCEL is solved and #2807 is merged

**Comment by romani** [comment]
@gtoison, is this issue fixed in 4.8.3.1 ?

**Comment by hazendaz** [comment]
No.  Spotbugs itself hasn't been released since 4.8.3.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by hazendaz** [comment]
Maven plugin receives updates as necessary for itself in between spotbugs releases.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by gtoison** [comment]
Hello @romani, this was fixed by #2807 (as well as #2819), right?

**Comment by romani** [comment]
Yes
